pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go

```diff
@@ -0,0 +1,246 @@
+			policy.GetName())
+	default:
+		return fmt.Errorf("unknown kind %q", kind)
+	}
+
+	if _, ok := policy.Object["description"]; ok {
+		scopedLog.Warn(warnTopLevelDescriptionField)
+		return ErrTopLevelDescriptionFound
+	}
+
+	policyBytes, err := policy.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	var filtered *unstructured.Unstructured
+
+	switch kind {
+	case cilium_v2.CNPKindDefinition:
+		cnp := new(cilium_v2.CiliumNetworkPolicy)
+		if err := json.Unmarshal(policyBytes, cnp); err != nil {
+			return err
+		}
+
+		var err error
+		filtered, err = toUnstructured(cnp, false)
+		if err != nil {
+			return err
+		}
+	case cilium_v2.CCNPKindDefinition:
+		ccnp := new(cilium_v2.CiliumClusterwideNetworkPolicy)
+		if err := json.Unmarshal(policyBytes, ccnp); err != nil {
+			return err
+		}
+
+		var err error
+		filtered, err = toUnstructured(ccnp, true)
+		if err != nil {
+			return err
+		}
+	default:
+		// We've already validated above that there can only be two kinds: CNP
+		// & CCNP. This is likely to be a developer error if hit, so panic.
+		scopedLog.WithField("kind", kind).Panic("Unexpected kind found when processing policy")
+	}
+
+	given, err := getFields(policy.Object)
+	if err != nil {
+		return err
+	}
+
+	expected, err := getFields(filtered.Object)
+	if err != nil {
+		return err
+	}
+
+	// Compare the expected policy with the given policy to find all unknown
+	// fields.
+	var r reporter
+	if !cmp.Equal(
+		expected,
+		given,
+		cmp.Reporter(&r),
+		cmpopts.SortSlices(func(o, n string) bool {
+			return o < n
+		}),
+	) {
+		scopedLog.Warn(warnUnknownFields)
+		return ErrUnknownFields{
+			extras: r.extras,
+		}
+	}
+
+	return nil
+}
+
+const (
+	warnTopLevelDescriptionField = "It seems you have a policy with a " +
+		"top-level description. This field is no longer supported. Please migrate " +
+		"your policy's description field under `spec` or `specs`."
+
+	warnUnknownFields = "It seems you have a policy with extra unknown fields. " +
+		"Consider removing these fields, as they have no effect. The presence " +
+		"of these fields may have introduced a false sense security, so please " +
+		"check whether your policy is actually behaving as you expect."
+)
+
+// ErrTopLevelDescriptionFound is the error returned if a policy contains a
+// top-level description field. Instead this field should be moved to under
+// (Rule).Description.
+var ErrTopLevelDescriptionFound = errors.New("top-level description field found")
+
+// ErrUnknownFields is an error representing the condition where unknown fields
+// were found within a policy during validation. Fields that are not expected
+// to be in the policy will be put inside the "extras" slice.
+type ErrUnknownFields struct {
+	extras []string
+}
+
+func (e ErrUnknownFields) Error() string {
+	return fmt.Sprintf("unknown fields found, extra:%v", e.extras)
+}
+
+func toUnstructured(obj interface{}, clusterwide bool) (*unstructured.Unstructured, error) {
```

Comment:
I know you already wrote this, but if you decide to use something ready-made, you might want to use `"k8s.io/apimachinery/pkg/runtime.DefaultUnstructuredConverter.ToUnstructured"`.

Reply:
ah! maybe this will solve the problem with marshaling / unmarshaling?

Reply:
Thanks! I wasn't aware of this library. It doesn't solve our marshalling / unmarshalling problem, but I did use it to avoid duplicating the `toStructured` functionality. I did end up finding a bug in the library which I've fixed in a fork. Not sure if it's worth keeping, but I've vendored the fork in case.